File: packages/sardis-sdk-python/src/sardis_sdk/resources/agents.py
"""
Agents resource for Sardis SDK.

This module provides both async and sync interfaces for managing agents.
Agents are the core identity entities in Sardis that can own wallets,
issue mandates, and be subject to policies.
"""
from __future__ import annotations

from typing import TYPE_CHECKING, Any, Dict, List, Optional, Union

from ..models import Agent, AgentCreate, AgentUpdate
from ..pagination import Page, PageInfo, create_page_from_response
from .base import AsyncBaseResource, SyncBaseResource

if TYPE_CHECKING:
    from ..client import AsyncSardisClient, SardisClient, TimeoutConfig


class AsyncAgentsResource(AsyncBaseResource):
    """Async resource for managing agents.

    Agents are the core identity entities in Sardis. They can:
    - Own wallets
    - Issue mandates
    - Be subject to policies

    Example:
        ```python
        async with AsyncSardisClient(api_key="...") as client:
            # Create an agent
            agent = await client.agents.create(
                name="my-agent",
                description="My AI agent",
            )

            # List all agents
            agents = await client.agents.list()

            # Get a specific agent
            agent = await client.agents.get("agent_123")

            # Update an agent
            agent = await client.agents.update(
                "agent_123",
                name="updated-name",
            )
        ```
    """

    async def create(
        self,
        name: str,
        description: Optional[str] = None,
        spending_limits: Optional[Dict[str, Any]] = None,
        policy: Optional[Dict[str, Any]] = None,
        metadata: Optional[Dict[str, Any]] = None,
        timeout: Optional[Union[float, "TimeoutConfig"]] = None,
    ) -> Agent:
        """Create a new agent.

        Args:
            name: Display name for the agent
            description: Optional description
            spending_limits: Dictionary of spending limits
            policy: Dictionary of policy rules
            metadata: Arbitrary key-value metadata
            timeout: Optional request timeout

        Returns:
            The created Agent object
        """
        payload: Dict[str, Any] = {"name": name}

        if description is not None:
            payload["description"] = description
        if spending_limits is not None:
            payload["spending_limits"] = spending_limits
        if policy is not None:
            payload["policy"] = policy
        if metadata is not None:
            payload["metadata"] = metadata

        data = await self._post("agents", payload, timeout=timeout)
        return Agent.model_validate(data)

    async def get(
        self,
        agent_id: str,
        timeout: Optional[Union[float, "TimeoutConfig"]] = None,
    ) -> Agent:
        """Get an agent by ID.

        Args:
            agent_id: The ID of the agent to retrieve
            timeout: Optional request timeout

        Returns:
            The Agent object
        """
        data = await self._get(f"agents/{agent_id}", timeout=timeout)
        return Agent.model_validate(data)

    async def list(
        self,
        limit: int = 100,
        offset: Optional[int] = None,
        timeout: Optional[Union[float, "TimeoutConfig"]] = None,
    ) -> List[Agent]:
        """List all agents.

        Args:
            limit: Maximum number of agents to return
            offset: Pagination offset
            timeout: Optional request timeout

        Returns:
            List of Agent objects
        """
        params: Dict[str, Any] = {"limit": limit}
        if offset is not None:
            params["offset"] = offset

        data = await self._get("agents", params=params, timeout=timeout)

        # Handle both list response and paginated response
        if isinstance(data, list):
            return [Agent.model_validate(item) for item in data]
        return [Agent.model_validate(item) for item in data.get("agents", data.get("items", []))]

    async def list_page(
        self,
        limit: int = 100,
        offset: int = 0,
        timeout: Optional[Union[float, "TimeoutConfig"]] = None,
    ) -> Page[Agent]:
        """List agents with pagination info.

        Args:
            limit: Maximum number of agents per page
            offset: Pagination offset
            timeout: Optional request timeout

        Returns:
            Page of Agent objects with pagination metadata
        """
        params: Dict[str, Any] = {"limit": limit, "offset": offset}
        data = await self._get("agents", params=params, timeout=timeout)

        return create_page_from_response(
            data=data,
            items_key="agents",
            item_parser=lambda x: Agent.model_validate(x),
        )

    async def update(
        self,
        agent_id: str,
        name: Optional[str] = None,
        description: Optional[str] = None,
        metadata: Optional[Dict[str, Any]] = None,
        is_active: Optional[bool] = None,
        timeout: Optional[Union[float, "TimeoutConfig"]] = None,
    ) -> Agent:
        """Update an agent.

        Args:
            agent_id: The ID of the agent to update
            name: New display name
            description: New description
            metadata: New metadata (merges with existing)
            is_active: Enable/disable the agent
            timeout: Optional request timeout

        Returns:
            The updated Agent object
        """
        payload: Dict[str, Any] = {}
        if name is not None:
            payload["name"] = name
        if description is not None:
            payload["description"] = description
        if metadata is not None:
            payload["metadata"] = metadata
        if is_active is not None:
            payload["is_active"] = is_active

        data = await self._patch(f"agents/{agent_id}", payload, timeout=timeout)
        return Agent.model_validate(data)

    async def delete(
        self,
        agent_id: str,
        timeout: Optional[Union[float, "TimeoutConfig"]] = None,
    ) -> None:
        """Delete an agent.

        Args:
            agent_id: The ID of the agent to delete
            timeout: Optional request timeout
        """
        await self._delete(f"agents/{agent_id}", timeout=timeout)


class AgentsResource(SyncBaseResource):
    """Sync resource for managing agents.

    Agents are the core identity entities in Sardis. They can:
    - Own wallets
    - Issue mandates
    - Be subject to policies

    Example:
        ```python
        with SardisClient(api_key="...") as client:
            # Create an agent
            agent = client.agents.create(
                name="my-agent",
                description="My AI agent",
            )

            # List all agents
            agents = client.agents.list()

            # Get a specific agent
            agent = client.agents.get("agent_123")

            # Update an agent
            agent = client.agents.update(
                "agent_123",
                name="updated-name",
            )
        ```
    """

    def create(
        self,
        name: str,
        description: Optional[str] = None,
        spending_limits: Optional[Dict[str, Any]] = None,
        policy: Optional[Dict[str, Any]] = None,
        metadata: Optional[Dict[str, Any]] = None,
        timeout: Optional[Union[float, "TimeoutConfig"]] = None,
    ) -> Agent:
        """Create a new agent.

        Args:
            name: Display name for the agent
            description: Optional description
            spending_limits: Dictionary of spending limits
            policy: Dictionary of policy rules
            metadata: Arbitrary key-value metadata
            timeout: Optional request timeout

        Returns:
            The created Agent object
        """
        payload: Dict[str, Any] = {"name": name}

        if description is not None:
            payload["description"] = description
        if spending_limits is not None:
            payload["spending_limits"] = spending_limits
        if policy is not None:
            payload["policy"] = policy
        if metadata is not None:
            payload["metadata"] = metadata

        data = self._post("agents", payload, timeout=timeout)
        return Agent.model_validate(data)

    def get(
        self,
        agent_id: str,
        timeout: Optional[Union[float, "TimeoutConfig"]] = None,
    ) -> Agent:
        """Get an agent by ID.

        Args:
            agent_id: The ID of the agent to retrieve
            timeout: Optional request timeout

        Returns:
            The Agent object
        """
        data = self._get(f"agents/{agent_id}", timeout=timeout)
        return Agent.model_validate(data)

    def list(
        self,
        limit: int = 100,
        offset: Optional[int] = None,
        timeout: Optional[Union[float, "TimeoutConfig"]] = None,
    ) -> List[Agent]:
        """List all agents.

        Args:
            limit: Maximum number of agents to return
            offset: Pagination offset
            timeout: Optional request timeout

        Returns:
            List of Agent objects
        """
        params: Dict[str, Any] = {"limit": limit}
        if offset is not None:
            params["offset"] = offset

        data = self._get("agents", params=params, timeout=timeout)

        # Handle both list response and paginated response
        if isinstance(data, list):
            return [Agent.model_validate(item) for item in data]
        return [Agent.model_validate(item) for item in data.get("agents", data.get("items", []))]

    def list_page(
        self,
        limit: int = 100,
        offset: int = 0,
        timeout: Optional[Union[float, "TimeoutConfig"]] = None,
    ) -> Page[Agent]:
        """List agents with pagination info.

        Args:
            limit: Maximum number of agents per page
            offset: Pagination offset
            timeout: Optional request timeout

        Returns:
            Page of Agent objects with pagination metadata
        """
        params: Dict[str, Any] = {"limit": limit, "offset": offset}
        data = self._get("agents", params=params, timeout=timeout)

        return create_page_from_response(
            data=data,
            items_key="agents",
            item_parser=lambda x: Agent.model_validate(x),
        )

    def update(
        self,
        agent_id: str,
        name: Optional[str] = None,
        description: Optional[str] = None,
        metadata: Optional[Dict[str, Any]] = None,
        is_active: Optional[bool] = None,
        timeout: Optional[Union[float, "TimeoutConfig"]] = None,
    ) -> Agent:
        """Update an agent.

        Args:
            agent_id: The ID of the agent to update
            name: New display name
            description: New description
            metadata: New metadata (merges with existing)
            is_active: Enable/disable the agent
            timeout: Optional request timeout

        Returns:
            The updated Agent object
        """
        payload: Dict[str, Any] = {}
        if name is not None:
            payload["name"] = name
        if description is not None:
            payload["description"] = description
        if metadata is not None:
            payload["metadata"] = metadata
        if is_active is not None:
            payload["is_active"] = is_active

        data = self._patch(f"agents/{agent_id}", payload, timeout=timeout)
        return Agent.model_validate(data)

    def delete(
        self,
        agent_id: str,
        timeout: Optional[Union[float, "TimeoutConfig"]] = None,
    ) -> None:
        """Delete an agent.

        Args:
            agent_id: The ID of the agent to delete
            timeout: Optional request timeout
        """
        self._delete(f"agents/{agent_id}", timeout=timeout)


__all__ = [
    "AsyncAgentsResource",
    "AgentsResource",
]

File: packages/sardis-sdk-js/src/resources/agents.ts
/**
 * Agents resource
 *
 * Agents are the core identity entities in Sardis. They can:
 * - Own wallets
 * - Issue mandates
 * - Be subject to spending policies
 */

import { BaseResource } from './base.js';
import type {
  Agent,
  CreateAgentInput,
  UpdateAgentInput,
  ListAgentsOptions,
  RequestOptions,
} from '../types.js';

export class AgentsResource extends BaseResource {
  private _normalize(agent: Agent): Agent {
    // Provide backwards-compatible alias: id := agent_id
    if (!agent.id) {
      (agent as unknown as { id: string }).id = agent.agent_id;
    }
    return agent;
  }

  /**
   * Create a new agent
   *
   * @example
   * ```typescript
   * const agent = await client.agents.create({
   *   name: 'Shopping Agent',
   *   description: 'Agent for e-commerce purchases',
   *   spending_limits: {
   *     per_transaction: '100.00',
   *     daily: '500.00',
   *   },
   * });
   * ```
   */
  async create(input: CreateAgentInput, options?: RequestOptions): Promise<Agent> {
    const agent = await this._post<Agent>('/api/v2/agents', input, options);
    return this._normalize(agent);
  }

  /**
   * Get an agent by ID
   *
   * @example
   * ```typescript
   * const agent = await client.agents.get('agent_abc123');
   * console.log(agent.name, agent.is_active);
   * ```
   */
  async get(agentId: string, options?: RequestOptions): Promise<Agent> {
    const agent = await this._get<Agent>(`/api/v2/agents/${agentId}`, undefined, options);
    return this._normalize(agent);
  }

  /**
   * List all agents
   *
   * @example
   * ```typescript
   * // List all agents
   * const agents = await client.agents.list();
   *
   * // List with pagination
   * const page = await client.agents.list({ limit: 10, offset: 20 });
   *
   * // List only active agents
   * const active = await client.agents.list({ is_active: true });
   * ```
   */
  async list(options?: ListAgentsOptions): Promise<Agent[]> {
    const params: Record<string, unknown> = {};

    if (options?.limit !== undefined) {
      params.limit = options.limit;
    }
    if (options?.offset !== undefined) {
      params.offset = options.offset;
    }
    if (options?.is_active !== undefined) {
      params.is_active = options.is_active;
    }

    const response = await this._get<{ agents: Agent[] } | Agent[]>(
      '/api/v2/agents',
      Object.keys(params).length > 0 ? params : undefined
    );

    // Handle both array and object response formats
    if (Array.isArray(response)) {
      return response.map((a) => this._normalize(a));
    }
    return (response.agents || []).map((a) => this._normalize(a));
  }

  /**
   * Update an agent
   *
   * @example
   * ```typescript
   * const updated = await client.agents.update('agent_abc123', {
   *   name: 'Updated Name',
   *   is_active: false,
   * });
   * ```
   */
  async update(agentId: string, input: UpdateAgentInput): Promise<Agent> {
    const agent = await this._patch<Agent>(`/api/v2/agents/${agentId}`, input);
    return this._normalize(agent);
  }

  /**
   * Delete an agent
   *
   * Note: This is a soft delete - the agent is deactivated, not removed.
   *
   * @example
   * ```typescript
   * await client.agents.delete('agent_abc123');
   * ```
   */
  async delete(agentId: string): Promise<void> {
    await this._delete(`/api/v2/agents/${agentId}`);
  }

  /**
   * Get an agent's associated wallet
   *
   * @example
   * ```typescript
   * const wallet = await client.agents.getWallet('agent_abc123');
   * console.log(wallet.addresses);
   * ```
   */
  async getWallet(agentId: string): Promise<{ wallet_id: string; addresses: Record<string, string> }> {
    return this._get(`/api/v2/agents/${agentId}/wallet`);
  }

  /**
   * Create a wallet for an agent
   *
   * @example
   * ```typescript
   * const wallet = await client.agents.createWallet('agent_abc123', {
   *   currency: 'USDC',
   *   limit_per_tx: '100.00',
   * });
   * ```
   */
  async createWallet(
    agentId: string,
    options?: {
      currency?: string;
      limit_per_tx?: string;
      limit_total?: string;
    }
  ): Promise<{ wallet_id: string; addresses: Record<string, string> }> {
    return this._post(`/api/v2/agents/${agentId}/wallet`, options || {});
  }
}
